crates/matrix-sdk-base/src/client.rs

```diff
@@ -1206,25 +1101,7 @@ impl BaseClient {
             new_rooms.invite.insert(room_id, new_info);
         }
 
-        // We're processing direct state events here separately
-        // because we want to have the push rules in place before we process
-        // rooms and their events, but we want to create the rooms before we
-        // process the `m.direct` account data event.
-        let has_new_direct_room_data = global_account_data_events
-            .iter()
-            .any(|event| event.event_type() == GlobalAccountDataEventType::Direct);
-        if has_new_direct_room_data {
-            self.process_direct_rooms(&global_account_data_events, &mut changes).await;
-        } else if let Ok(Some(direct_account_data)) =
-            self.store.get_account_data_event(GlobalAccountDataEventType::Direct).await
-        {
-            debug!("Found direct room data in the Store, applying it");
-            if let Ok(direct_account_data) = direct_account_data.deserialize() {
-                self.process_direct_rooms(&[direct_account_data], &mut changes).await;
-            } else {
-                warn!("Failed to deserialize direct room account data");
-            }
-        }
+        account_data.apply(&mut changes, &self.store).await;
```

Comment:
Interface wise it's a bit weird for a processor to be stateful. I would've imagined it a bit more functional with the `Processor` only having static methods on itself and producing various versions of `AccountData` on successive calls
e.g. `Processor::process(Response) -> AccountData`; `Processor::process(StateChanges, AccountData) -> AccountData`

Reply:
To expand on this. We introduced a `Processor` concept for sliding sync responses as well, but it also gets mutated and it needs to be mutated in the correct order, otherwise it panics:

https://github.com/matrix-org/matrix-rust-sdk/blob/59d3608c32c94b7bf210a9bee1ba2733d37b2cec/crates/matrix-sdk/src/sliding_sync/client.rs#L288-L290

A good way to handle this sort of requirements might be to introduce typestates, i.e. like @stefanceriu suggest, but each `process()` function accepts a certain part of the response and returns a different processor.

For the `SlidingSyncResponseProcessor` this would look somewhat like this:


```rust

struct SlidingSyncEncryptionProcessor {
...
}

impl SlidingSyncEncryptionProcessor {
    fn process(self, extensions: &Extensions) -> SlidingSyncRoomProcessor {
    }
}

struct SlidingSyncRoomProcessor {
    ...
}

...
```

This avoids the need for any sort of asserts to ensure things happen in the right order, because the right order is encoded in the types.

Reply:
Yeah, I can definitely live with that, it's already a great step forward. 
Another quick win might be to assert that it's consumed once and only once internally after moving the data, or perhaps even renaming `apply` to `transfer_into` or something.

Reply:
For the sake of completeness, forgot to respond here, so responded in private after a ping from stefan:

> this doesn't buy us anything because the type is must_use (so must be consumed at least once) and since apply(self) consumes self, it's consumed at most once
> static typing ftw!